### PR TITLE
audit(github): prefer `/archive/refs/tags` urls over `/archive`

### PR DESCRIPTION
**Pull request by chenrui333**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

followup https://github.com/mislav/bump-homebrew-formula-action/pull/53

seeing couple of issues recently in homebrew-core for the artifact urls

### CODE REVIEW

**Comment by Bo98** [comment]
Good idea, but this won't work with commit URLs like with LuaJIT so you will need to adjust the regex to exclude those types of URLs.

**Comment by chenrui333** [comment]
yeah, I will update all the urls later :)

**Comment by chenrui333** [comment]
Added a big PR to fix all the url formats, https://github.com/Homebrew/homebrew-core/pull/152092

**Comment by chenrui333** [comment]
should be good to go

File: Library/Homebrew/rubocops/urls.rb [hunk 1]
```diff
@@ -217,6 +217,13 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
             problem "Use /archive/ URLs for GitHub tarballs (url is #{url})."
           end
 
+          archive_ref_tags_gh_pattern = %r{https://.*github.*/archive/(?![a-fA-F0-9]{40})(?!refs/tags/).*\.tar\.gz$}
+          audit_urls(urls, archive_ref_tags_gh_pattern) do |_, url|
+            next if url.end_with?(".git")
+
+            problem "Use /archive/refs/tags URLs for GitHub tarballs (url is #{url})."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/urls.rb:224]: Would be nice to write an autocorrect for this if possible?

Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/urls.rb:224]: Can these both trigger at once?

Comment (chenrui333) [reply Library/Homebrew/rubocops/urls.rb:224]: any example of doing autocorrect?

Comment (chenrui333) [reply Library/Homebrew/rubocops/urls.rb:224]: i assume you are talking about the `archive_gh_pattern`, i just tested, looks like it cannot be triggered all together

Comment (MikeMcQuaid) [reply Library/Homebrew/rubocops/urls.rb:224]: Cool, thanks 👍🏻

Comment (MikeMcQuaid) [reply Library/Homebrew/rubocops/urls.rb:224]: Perhaps also worth checking it won't suggest a change to a format rejected by this audit.

Comment (MikeMcQuaid) [reply Library/Homebrew/rubocops/urls.rb:224]: https://github.com/Homebrew/brew/blob/121fc592cf4452040ffd877299ecfbb2b4fe86aa/Library/Homebrew/rubocops/homepage.rb#L46

Can wait till a follow-up PR but, if it's not too hard: autocorrects are always nice to have.

Comment (chenrui333) [reply Library/Homebrew/rubocops/urls.rb:224]: oh, i see your point, that is right. let me add a fix

Comment (chenrui333) [reply Library/Homebrew/rubocops/urls.rb:224]: actually could be a followup pr to autocorrect the fix.

Comment (chenrui333) [reply Library/Homebrew/rubocops/urls.rb:224]: 👍

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by chenrui333** [comment]
The current audit failures are due to the repo removal issues for `hashpump` and `nethacked`
```
/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/h/hashpump.rb:4:3: C: FormulaAudit/Urls: Use /archive/refs/tags URLs for GitHub tarballs (url is https://github.com/bwall/HashPump/archive/v1.2.0.tar.gz).
  url "https://github.com/bwall/HashPump/archive/v1.2.0.tar.gz"
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/n/nethacked.rb:24:3: C: FormulaAudit/Urls: Use /archive/refs/tags URLs for GitHub tarballs (url is https://github.com/nethacked/nethacked/archive/1.0.tar.gz).
  url "https://github.com/nethacked/nethacked/archive/1.0.tar.gz"
```

**Comment by Bo98** [comment]
You will need to change them still. Can do it in separate PRs to make it obvious it's not tested (though the orginal URL is broken anyway so it doesn't matter much).

**Comment by chenrui333** [comment]
I see, let me update those two then.

**Comment by chenrui333** [comment]
Should be good now.

**Comment by MikeMcQuaid** [comment]
I'm happy to merge when @Bo98 is ✅

**Review by Bo98 — APPROVED** [review]

**Comment by mahrud** [comment]
This PR broke using a branch archive for head builds. What I mean is that this:
```ruby
  head do
    url "https://github.com/Macaulay2/M2/archive/refs/heads/master.tar.gz"
  end
```
Now gives an error:
```ruby
Formula/macaulay2.rb:24:5: C: FormulaAudit/Urls: Use /archive/refs/tags URLs for GitHub tarballs (url is https://github.com/Macaulay2/M2/archive/refs/heads/master.tar.gz).
    url "https://github.com/Macaulay2/M2/archive/refs/heads/master.tar.gz"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
This is only relevant for `--head` builds, where I think it's important to allow getting the latest commits from a branch before they are tagged.

**Comment by Bo98** [comment]
Yeah we should update the regex to exclude `refs/heads` - that makes sense to me.

**Comment by MikeMcQuaid** [comment]
> Yeah we should update the regex to exclude `refs/heads` - that makes sense to me.

@chenrui333 can you do this? Thanks!

**Comment by mahrud** [comment]
I opened #16155 to address this for now.

**Pull request merged**